tempesta_fw/sched/tfw_sched_hash.c

```diff
@@ -245,12 +200,114 @@ tfw_sched_hash_get_srv_conn(TfwMsg *msg, TfwServer *srv)
 	return NULL;
 }
 
+static void
+tfw_sched_hash_cleanup(TfwSrvGroup *sg)
+{
+	size_t si;
+	TfwHashSrvList *sl = sg->sched_data;
+
+	if (!sl)
+		return;
+
+	for (si = 0; si < sl->srv_n; ++si) {
+		if (sl->srvs[si].conn)
+			kfree(sl->srvs[si].conn);
+		if (sl->srvs[si].hash)
+			kfree(sl->srvs[si].hash);
+	}
+
+	kfree(sl);
+	sg->sched_data = NULL;
+}
+
+static void
+tfw_sched_hash_del_grp(TfwSrvGroup *sg)
+{
+	tfw_sched_hash_cleanup(sg);
+}
+
+/**
+ * Validate the integrity of a group.
+ *
+ * Make sure that number of servers in the group, and the number
+ * of connections for each server match the recorded values.
+ */
+static int
+tfw_sched_hash_validate_grp(TfwSrvGroup *sg)
+{
+	size_t si = 0, ci;
+	TfwServer *srv;
+	TfwSrvConn *srv_conn;
+
+	list_for_each_entry(srv, &sg->srv_list, list) {
+		ci = 0;
+		list_for_each_entry(srv_conn, &srv->conn_list, list)
+			++ci;
+		if (ci > srv->conn_n)
+			return -EINVAL;
+		++si;
+	}
+	if (si > sg->srv_n)
+		return -EINVAL;
+
+	return 0;
+}
+
+static int
+tfw_sched_hash_add_grp(TfwSrvGroup *sg)
+{
+	int ret = -ENOMEM;
+	size_t size, ci;
+	unsigned int sum_conn_n;
+	TfwServer *srv;
+	TfwSrvConn *srv_conn;
+	TfwHashSrv *hsrv;
+	TfwHashSrvList *sl;
+
+	if (tfw_sched_hash_validate_grp(sg))
+		return -EINVAL;
```

Comment:
At the below you do loop with relatively heavy body, so additional check for two integers (index and the size of the array) won't hurt performance. We always keep loop bodies as small as possible and move conditions outside of the loop, but only if the condition hasn't been checked for _each_ iteration. There is absolutely no sense in splitting a loop to two loops.